Sources/Sentry/Profiling/SentryProfileTimeseries.mm

```diff
@@ -0,0 +1,92 @@
+#import "SentryProfileTimeseries.h"
+#import "SentryEvent+Private.h"
+#import "SentryInternalDefines.h"
+#import "SentryLog.h"
+#import "SentryTransaction.h"
+
+std::mutex _gSamplesArrayLock;
+
+/**
+ * Print a debug log to help diagnose slicing errors.
+ * @param start @c YES if this is an attempt to find the start of the sliced data based on the
+ * transaction start; @c NO if it's trying to find the end of the sliced data based on the
+ * transaction's end, to accurately describe what's happening in the log statement.
+ */
+void
+logSlicingFailureWithArray(
+    NSArray<SentrySample *> *array, SentryTransaction *transaction, BOOL start)
+{
+    if (!SENTRY_CASSERT(array.count > 0, @"Should not have attempted to slice an empty array.")) {
+        return;
+    }
+
+    if (![SentryLog willLogAtLevel:kSentryLevelDebug]) {
+        return;
+    }
+
+    const auto firstSampleAbsoluteTime = array.firstObject.absoluteTimestamp;
+    const auto lastSampleAbsoluteTime = array.lastObject.absoluteTimestamp;
+    const auto firstSampleRelativeToTransactionStart
+        = firstSampleAbsoluteTime - transaction.startSystemTime;
+    const auto lastSampleRelativeToTransactionStart
+        = lastSampleAbsoluteTime - transaction.startSystemTime;
+    SENTRY_LOG_DEBUG(@"[slice %@] Could not find any%@ sample taken during the transaction "
+                     @"(first sample taken at: %llu; last: %llu; transaction start: %llu; end: "
+                     @"%llu; first sample relative to transaction start: %lld; last: %lld).",
+        start ? @"start" : @"end", start ? @"" : @" other", firstSampleAbsoluteTime,
+        lastSampleAbsoluteTime, transaction.startSystemTime, transaction.endSystemTime,
+        firstSampleRelativeToTransactionStart, lastSampleRelativeToTransactionStart);
+}
+
+NSArray<SentrySample *> *_Nullable slicedProfileSamples(
+    NSArray<SentrySample *> *samples, SentryTransaction *transaction)
+{
+    NSArray<SentrySample *> *samplesCopy;
+    {
+        std::lock_guard<std::mutex> l(_gSamplesArrayLock);
+        samplesCopy = [samples copy];
```

Comment:
`h`: This fixes the problem, but adding a similar bug in the future is likely. I think we might still have a bug in `serializedSamplesWithRelativeTimestamps` as it also iterates the samples array, but it doesn't use a copy or a sync mechanism.

I would prefer moving this back to the profiler to have all locks in one file, or we write a class that hides the samples array and ensures thread-safe access.

Reply:
Okay, then we can refactor this later if needed. Most importantly, we ship a fix quickly.